### PR TITLE
Adds jumping

### PR DESCRIPTION
**Pull request by Lumipharon**
## About The Pull Request
Adds a jumping mechanic.

https://github.com/tgstation/TerraGov-Marine-Corps/assets/7869430/4e9662d3-cf05-4364-98d5-9e17b7a70f48

Press button, you jump. Allows you to leap over certain obstacles such as tables and window frames instead of doing an annoying and slow click drag.

In light of this, platforms have been changed back to dense. No more slowdown shananigans, you just jump over them (climbing still works if needed).

Xenos can also jump, however they have a longer cooldown due to no stamina.

https://github.com/tgstation/TerraGov-Marine-Corps/assets/7869430/cf7c42b4-690a-41ab-b82d-90803bca3fb5

https://github.com/tgstation/TerraGov-Marine-Corps/assets/7869430/e2de2e0a-52b0-4f9f-a808-880adf3d8cef

Jump functionality is very modular and can be applied to any mob, with its exact effects customised. For example you can add jump sound effects, projectile or mob pass etc.

## Why It's Good For The Game
Jumping is cool. Click drag climbing is lame.
Adds exciting new features and gameplay opportunities.
## Changelog
:cl:
add: Adds a jump mechanic, to leap over obstacles like window frames
balance: Platforms are dense again, in light of the ability to jump over them
/:cl:

### CODE REVIEW

**Review by spookydonut — APPROVED** [review]

**Comment by Tonklaatech** [comment]
SPEEEEED MERGEEEEE

**Comment by bluezorua** [comment]
unironically add this as some kind of armour module or something

**Comment by FosterGrime-Heymynameslouie** [comment]
should be turned into a req module ngl

**Comment by SilencedMP5A5** [comment]
speedmerge

**Comment by mc-oofert** [comment]
this should be a module

**Comment by Mrrpip** [comment]
...So....Are we giving Vordt's theme to Crusher or to Behemoth?

**Comment by Pariah919** [comment]
This has HIGH sickness. Merge this NOW.

File: code/datums/elements/jump.dm [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+/datum/element/jump/proc/do_jump(mob/living/jumper)
+	SIGNAL_HANDLER
+	if(jumper.incapacitated(TRUE))
+		return
+
+	if(TIMER_COOLDOWN_CHECK(jumper, JUMP_ELEMENT_COOLDOWN))
+		return
+
+	if((jumper.getStaminaLoss()) > -stamina_cost)
+		to_chat(jumper, span_warning("Catch your breathe!"))
```
Comment (ivanmixo) [inline code/datums/elements/jump.dm:39]:
```suggestion
		to_chat(jumper, span_warning("Catch your breath!"))
```
Also LAME let people kill themselves by jumping too much

**Comment by Muffindrake** [comment]
speedmerge this now

**Comment by lKiseki** [comment]
I thought this would be giga broken but it's actually unironically not so bad. Just wish the jump was more realistic instead of this Looney Tunes thing.

**Comment by AlexCrimson** [comment]
I will be the anti-FUN and complain about the fact the sound is annoying. Then the super serious balance implications of letting marines jump their way out of harm. Like a Xeno surround, or if they overextend through barricades. 

Give it to Xenos so at least they can prove how silly it is to let people jump over barricades.

You can jump over fucking lava

grumble grumble

**Comment by Kapu1178** [comment]
This is funny but extremely broken. Marines can jump over xenos to get out of bodyblocking after they fuck up.

Also Marines just win on Magmoor for free.

**Comment by Aporhtonoma** [comment]
agility 100

**Comment by Pariah919** [comment]
> I will be the anti-FUN and complain about the fact the sound is annoying. Then the super serious balance implications of letting marines jump their way out of harm. Like a Xeno surround, or if they overextend through barricades.
> 
> Give it to Xenos so at least they can prove how silly it is to let people jump over barricades.
> 
> You can jump over fucking lava
> 
> grumble grumble

xenos can't jump because they leap

**Comment by ClosetedSkeletonGH** [comment]
Could very easily be req moon boots or a parkour button (why does hopping over a windowsill take longer in a video game than in IRL for marines (also xenos take forever too when theyre supposed to be pretty acrobatic?))

**Comment by Krypandenej** [comment]
Hear me out, sign this to whatever req module, like an armor module or special boots, and make them hackable with a multitool for the OG goofy noise and SUPERFLIP

**Comment by Beepinex** [comment]
please make this into a jump-jet module 🥺

**Comment by yyzsong** [comment]
> This should be a module this should be a module!

It is. Just go use jetpack.
Giving every marine jetpack-lite sucks. Bodyblocking is no longer and thing and sometimes they can dodge abilities by jumping.

**Comment by thegleendoctor** [comment]
look, i literally made a new GIT account after 10 years to comment this. I get all SS13 servers have for some reason abandoned all sense of sound design, aside from _the server_ that is definitely not on GITHUB. 

When I'm playing I want to hear the distant roar of gunfire, the cries of marines making a desperate last push, reverberating explosions of a devastating OB, not the bargain bin looney toons sound library dumped every 15 seconds, it's as common as GUNFIRE in any given combat engagement, it OVERPOWERS it. 

Now, I understand this server gave up on t any semblance of the med RP thing ages ago and it's essentially a TF2 lobby, so bringing up 'muh immersion' isn't gonna be heard by anyone, but this is the official shark jump in regards to goofy ahh mechanics. Pun intended. 

This isn't even TOUCHING the 'balance' this merge provides, just the horrid sound design. ADD A GAME SETTING TICK TO **MUTE** **_THIS_**

File: code/datums/elements/jump.dm [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+	if((jumper.getStaminaLoss()) > -stamina_cost)
+		to_chat(jumper, span_warning("Catch your breath!"))
+		return
+
+	playsound(jumper, "jump", 65)
+	jumper.layer = ABOVE_MOB_LAYER
+
+	jumper.adjustStaminaLoss(stamina_cost)
+	jumper.flags_pass |= HOVERING|PASSPROJECTILE
```
Comment (Kapu1178) [inline code/datums/elements/jump.dm:46]: This is obscenely broken and regularly fucks over Xenos like Boiler.

File: code/datums/elements/jump.dm [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+	jumper.animation_spin(5, 2, jumper.dir == WEST ? FALSE : TRUE)
+	animate(jumper, pixel_y = jumper.pixel_y + 40, layer = ABOVE_MOB_LAYER, time = jump_duration / 2, easing = CIRCULAR_EASING|EASE_OUT, flags = ANIMATION_PARALLEL)
+	animate(pixel_y = initial(jumper.pixel_y), time = jump_duration / 2, easing = CIRCULAR_EASING|EASE_IN)
```
Comment (Kapu1178) [inline code/datums/elements/jump.dm:50]: It is not fun to not be able to tell what tile a marine is on while trying to bump attack them, let alone try to click on them with an ability.

File: code/game/sound.dm [hunk 4]
```diff
@@ -333,6 +333,8 @@ A good representation is: 'byond applies a volume reduction to the sound every X
 			S = pick("sound/voice/human_male_preburst1.ogg", 'sound/voice/human_male_preburst2.ogg', 'sound/voice/human_male_preburst3.ogg', 'sound/voice/human_male_preburst4.ogg', 'sound/voice/human_male_preburst5.ogg', 'sound/voice/human_male_preburst6.ogg', 'sound/voice/human_male_preburst7.ogg', 'sound/voice/human_male_preburst8.ogg', 'sound/voice/human_male_preburst9.ogg', 'sound/voice/human_male_preburst10.ogg')
 		if("female_preburst")
 			S = pick("sound/voice/human_female_preburst1.ogg", 'sound/voice/human_female_preburst2.ogg', 'sound/voice/human_female_preburst3.ogg')
+		if("jump")
+			S = pick('sound/effects/bounce_1.ogg','sound/effects/bounce_2.ogg','sound/effects/bounce_3.ogg','sound/effects/bounce_4.ogg')
```
Comment (Kapu1178) [inline code/game/sound.dm:337]: I already started muting my game, this sound is grating.

**Comment by Joe13413** [comment]
do you mfers seriously expect this change to stay in for more than a day
like you bros be talking about this like it isnt some late april fools joke

it might get added in some capacity but with less sillyness so calm down about the sillyness you fun haters

**Comment by mc-oofert** [comment]
> When I'm playing I want to hear the distant roar of gunfire, the cries of marines making a desperate last push, reverberating explosions of a devastating OB, not the bargain bin looney toons sound library dumped every 15 seconds, it's as common as GUNFIRE in any given combat engagement, it OVERPOWERS it.

this is TDM not some goofy ahh battlefield 4
+ title obviously implies joke

File: code/datums/elements/jump.dm (hunk 2 above)
Comment (mc-oofert) [reply code/datums/elements/jump.dm:46]: skill issue

**Comment by mc-oofert** [comment]
the nerf is real

**Comment by Lumipharon** [comment]
Sorry bros, too many downvotes, I had to cave to the pressure.

**Comment by Kapu1178** [comment]
Now that this officially isn't a joke, let's talk numbers.

Giving Marines the ability to jump over fire and *cades* is an undeniable massive buff. Xenos are already struggling a bit with flamer and WP spam, this just makes it worse.

The worst part is the Marine winrate is currently a [comfortable 60%](https://statbus.psykzz.com/api/winrate?delta=14). I don't think in good faith this can be justified with the way the game is right now, it's far too forgiving to marines who fuck up, and allows for even more aggressive plays with fire, something that is already extremely strong right now due to Hold Order.

**Comment by Wayland-Smithy** [comment]
> In light of this, I'll probably change platforms back to dense.

My brain is so broken by this PR I don't even know if this part is a troll or not.
Is it?

**Comment by Lumipharon** [comment]
> > In light of this, I'll probably change platforms back to dense.
> 
> My brain is so broken by this PR I don't even know if this part is a troll or not. Is it?
> 
> Edit: you can't make this up... it was codified as I posted that. Wow what a bad move calling climbing CBT then forcing it on anyone without the "jump" i.e xenos all the time (I know they technically can, but slower if allowed at all which I doubt). 👎

Wuh? I'm not sure what you're saying, but xenos can jump over platforms exactly the same as humans. Its literally instant, so a massive improvement over both the current slowdown system and the old clickdrag climb system.

**Comment by Burger1998** [comment]
> look, i literally made a new GIT account after 10 years to comment this. I get all SS13 servers have for some reason abandoned all sense of sound design, aside from _the server_ that is definitely not on GITHUB... _BUT_
> 
> When I'm playing I want to hear the distant roar of gunfire, the cries of marines making a desperate last push, reverberating explosions of a devastating OB, not the bargain bin looney toons sound library dumped every 15 seconds, it's as common as GUNFIRE in any given combat engagement, it OVERPOWERS it.
> 
> Now, I understand this server gave up on any semblance of the med RP thing ages ago and it's essentially a TF2 lobby, so bringing up 'muh immersion' isn't gonna be heard by anyone, but this is the official shark jump in regards to goofy ahh mechanics. Pun intended.
> 
> This isn't even TOUCHING the 'balance' this merge provides, just the horrid sound design. ADD A GAME SETTING TICK TO **MUTE** **_THIS_**

this is very likely just a flavor of the month meme pr so i doubt we're getting settings for it.(the sound is incredibly funny and everyone who mutes it should be locked up)

**Comment by 1kajowsmol** [comment]
this pr sucks.

the sound is annoying, and the animation itself is goofy as hell even for tgmc standards

**Comment by Wayland-Smithy** [comment]
> Wuh? I'm not sure what you're saying, but xenos can jump over platforms exactly the same as humans. Its literally instant, so a massive improvement over both the current slowdown system and the old clickdrag climb system.

Okay I pulled the pr and tested locally so I see now it's added an init, so that is cool actually. 2 secs might be on the longer side, seems okay with no TD.

However platforms seem to not allow climbing by xenos via any other method now. Please consider keeping the old slowdown with this just added on top, or probably better to make bumping into them automatically start a climb. I think they did that before the slowdown?

**Comment by yyzsong** [comment]
> probably better to make bumping into them automatically start a climb.

Fuck no. Someone's tried that before and it was terrible. I don't ever wanna see it again.

Platform issue is mostly just them being unslashable. Fix that, if anything.

**Comment by Burger1998** [comment]
is this a meme pr or not im scared now

**Comment by Aporhtonoma** [comment]
we should propably restrict cade jumping, mob jumping, bullet dodging and keep it to only jumping over window frames.Or do it as other codebases have done similar stuff and have people stunned if they jump onto someone.Its fun and for the sake of fun it should be left as is... but the people are coping.

**Comment by Aidenzack1** [comment]
Hey, incase this wasn't considered (doesn't seem like it), having high pitched noises play ever 5 seconds is really bad for people who are sensitive to high pitches noises! Shocker, I actually can't play TGMC without getting a bad headache, so please change the sound, make it toggleable, ideally both. Repetitive intrusive sounds are in general, terrible ideas. Other than that, I have my negative opinions on the balance of this, but it's nothing that hasn't already been said.

**Review by BraveMole — APPROVED** [review]

**Comment by Overdrive659** [comment]
This PR will completely break Magmoor, more than it already is B)

**Review by spookydonut — APPROVED** [review]

**Comment by timothymtorres** [comment]
Platform mechanics are cringe. This makes TGMC feel like Super Mario Bros.

**Pull request merged**